dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/zorder/InsertZorderBeforeWriting.scala

```
@@ -0,0 +1,155 @@
+ */
+
+package org.apache.kyuubi.sql.zorder
+
+import java.util.Locale
+
+import org.apache.spark.sql.SparkSession
+import org.apache.spark.sql.catalyst.catalog.CatalogTable
+import org.apache.spark.sql.catalyst.expressions.{Ascending, NullsLast, SortOrder}
+import org.apache.spark.sql.catalyst.plans.logical.{LogicalPlan, Project, Repartition, RepartitionByExpression, Sort}
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.execution.command.CreateDataSourceTableAsSelectCommand
+import org.apache.spark.sql.execution.datasources.InsertIntoHadoopFsRelationCommand
+import org.apache.spark.sql.hive.execution.{CreateHiveTableAsSelectCommand, InsertIntoHiveTable, OptimizedCreateHiveTableAsSelectCommand}
+
+import org.apache.kyuubi.sql.KyuubiSQLConf
+
+/**
+ * TODO: shall we forbid zorder if it's dynamic partition inserts ?
+ * Insert zorder before writing datasource if the target table properties has zorder properties
+ */
+case class InsertZorderBeforeWritingDatasource(session: SparkSession)
+  extends InsertZorderHelper {
+  override def applyInternal(plan: LogicalPlan): LogicalPlan = plan match {
+    case insert: InsertIntoHadoopFsRelationCommand if insert.query.resolved &&
+      insert.bucketSpec.isEmpty && insert.catalogTable.isDefined &&
+          isZorderEnabled(insert.catalogTable.get.properties) =>
+      val newQuery = insertZorder(insert.catalogTable.get, insert.query)
+      if (newQuery.eq(insert.query)) {
+        insert
+      } else {
+        insert.copy(query = newQuery)
+      }
+
+    case ctas: CreateDataSourceTableAsSelectCommand if ctas.query.resolved &&
+      ctas.table.bucketSpec.isEmpty && isZorderEnabled(ctas.table.properties) =>
+      val newQuery = insertZorder(ctas.table, ctas.query)
+      if (newQuery.eq(ctas.query)) {
+        ctas
+      } else {
+        ctas.copy(query = newQuery)
+      }
+
+    case _ => plan
+  }
+}
+
+/**
+ * TODO: shall we forbid zorder if it's dynamic partition inserts ?
```

I think we can throw an exception when table partition columns are specified as zorder columns since it makes no sense.